DatadogSessionReplay/Sources/DatadogCoreIntegration/SRContextPublisher.swift

```diff
@@ -9,17 +9,38 @@ import Datadog
 
 /// Publisher that sets Session Replay context for being utilized by other Features.
 internal class SRContextPublisher {
+    internal struct SRContext {
+        var hasReplay: Bool = false
+        var recordsCount: [String: Int64] = [:]
+    }
+
     private weak var core: DatadogCoreProtocol?
 
+    private var srContext = SRContext() {
+        didSet {
+            core?.set(
+                feature: RUMDependency.srBaggageKey,
+                attributes: {
+                    [
+                        RUMDependency.hasReplay: self.srContext.hasReplay,
+                        RUMDependency.recordsCount: self.srContext.recordsCount
+                    ]
+                }
+            )
+        }
+    }
+
     init(core: DatadogCoreProtocol) {
         self.core = core
     }
 
     /// Notifies other Features on the state of  Session Replay recording.
-    func setRecordingIsPending(_ value: Bool) {
-        core?.set(
-            feature: RUMDependency.srBaggageKey,
-            attributes: { [RUMDependency.hasReplay: value] }
-        )
+    func setHasReplay(_ value: Bool) {
+        srContext.hasReplay = value
+    }
+
+    /// Notifies other Features on the state of  Session Replay recording.
+    func setRecordsCount(_ value: [String: Int64]) {
+        srContext.recordsCount = value
     }
```

Comment:
The `srContext` is a mutable state accessed from 3 different threads:
* SDK's Core "context thread" in `set(attributes: {})`
* SR's processor thread in `setRecordsCount(_:)`
* SR's recorder recorder (main) thread in `setHasReplay(_)`

Not synchronised access causes data races and obscure crashes. In this exact case, it is easy to reproduce it in demo app by setting recording frequency to `0.01s` and playing with the app for few seconds with TSAN enabled:
```
==================
WARNING: ThreadSanitizer: Swift access race (pid=95630)
  Modifying access of Swift variable at 0x00011a8ba860 by thread T11:
    #0 SRContextPublisher.srContext.modify <null>:406867456 (SRSampleApp:arm64+0x1000416d0) (BuildId: de3414cd17d636aaa05df4dfc3a6f9ce32000000200000000700000000000f00)
    #1 SRContextPublisher.setRecordsCount(_:) <null>:406867456 (SRSampleApp:arm64+0x100041a98) (BuildId: de3414cd17d636aaa05df4dfc3a6f9ce32000000200000000700000000000f00)
    #2 Processor.trackRecord(key:value:) <null>:406867456 (SRSampleApp:arm64+0x10005c95c) (BuildId: de3414cd17d636aaa05df4dfc3a6f9ce32000000200000000700000000000f00)
    ...

  Previous read of size 1 at 0x00011a8ba860 by thread T5:
    #0 SRContextPublisher.srContext.getter <null>:406867456 (SRSampleApp:arm64+0x1000414f0) (BuildId: de3414cd17d636aaa05df4dfc3a6f9ce32000000200000000700000000000f00)
    #1 closure #1 in SRContextPublisher.srContext.didset <null>:406867456 (SRSampleApp:arm64+0x100041330) (BuildId: de3414cd17d636aaa05df4dfc3a6f9ce32000000200000000700000000000f00)
    ...

  Location is heap block of size 48 at 0x00011a8ba840 allocated by main thread:
    #0 __sanitizer_mz_malloc <null>:408965280 (libclang_rt.tsan_iossim_dynamic.dylib:arm64+0x5e618) (BuildId: ce023e0a240933eaa69ce3fe788cd4a932000000200000000700000000000e00)
    #1 _malloc_zone_malloc_instrumented_or_legacy <null>:404773952 (libsystem_malloc.dylib:arm64+0x18278) (BuildId: 48ffcd7223e73c058124f01a518973ae32000000200000000700000000021000)
    #2 SessionReplayFeature.init(core:configuration:) <null>:406867456 (SRSampleApp:arm64+0x100044bd0) (BuildId: de3414cd17d636aaa05df4dfc3a6f9ce32000000200000000700000000000f00)
    ...

  Thread T11 (tid=1012357, running) is a GCD worker thread

  Thread T5 (tid=1012315, running) is a GCD worker thread

SUMMARY: ThreadSanitizer: Swift access race (SRSampleApp:arm64+0x1000416d0) (BuildId: de3414cd17d636aaa05df4dfc3a6f9ce32000000200000000700000000000f00) in SRContextPublisher.srContext.modify+0x7c
==================
```

Reply:
That's a great catch. Haven't really considered threading in this case, was really looking for a quick win.

We discussed with @maxep that we potentially want change the implementation of the core.set() to merge FeatureBaggages instead of overriding them, I was afraid about the impact though.

I'll revisit this with threading in mind and we can consider for core change in the near future.